File: server/src/main/java/se/kth/id2203/bootstrapping/BootstrapClient.java
/*
 * The MIT License
 *
 * Copyright 2017 Lars Kroll <lkroll@kth.se>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package se.kth.id2203.bootstrapping;

import java.util.UUID;
import org.slf4j.LoggerFactory;
import se.kth.id2203.networking.Message;
import se.kth.id2203.networking.NetAddress;
import se.sics.kompics.ClassMatchedHandler;
import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Handler;
import se.sics.kompics.Negative;
import se.sics.kompics.Positive;
import se.sics.kompics.Start;
import se.sics.kompics.network.Network;
import se.sics.kompics.timer.CancelPeriodicTimeout;
import se.sics.kompics.timer.SchedulePeriodicTimeout;
import se.sics.kompics.timer.Timer;

/**
 *
 * @author Lars Kroll <lkroll@kth.se>
 */
public class BootstrapClient extends ComponentDefinition {

    public static enum State {

        WAITING, STARTED;
    }

    private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(BootstrapClient.class);
    //******* Ports ******
    final Negative<Bootstrapping> bootstrap = provides(Bootstrapping.class);
    final Positive<Timer> timer = requires(Timer.class);
    final Positive<Network> net = requires(Network.class);
    //******* Fields ******
    private final NetAddress self = config().getValue("id2203.project.address", NetAddress.class);
    private final NetAddress server = config().getValue("id2203.project.bootstrap-address", NetAddress.class);

    private State state = State.WAITING;

    private UUID timeoutId;

    //******* Handlers ******
    protected final Handler<Start> startHandler = new Handler<Start>() {

        @Override
        public void handle(Start event) {
            LOG.debug("Starting bootstrap client on {}", self);
            long timeout = (config().getValue("id2203.project.keepAlivePeriod", Long.class) * 2);
            SchedulePeriodicTimeout spt = new SchedulePeriodicTimeout(timeout, timeout);
            spt.setTimeoutEvent(new BSTimeout(spt));
            trigger(spt, timer);
            timeoutId = spt.getTimeoutEvent().getTimeoutId();
        }
    };
    protected final Handler<BSTimeout> timeoutHandler = new Handler<BSTimeout>() {

        @Override
        public void handle(BSTimeout e) {

            if (state == State.WAITING) {
                trigger(new Message(self, server, CheckIn.event), net);
            } else if (state == State.STARTED) {
                trigger(new Message(self, server, Ready.event), net);
                suicide();
            }
        }
    };

    protected final ClassMatchedHandler<Boot, Message> bootHandler = new ClassMatchedHandler<Boot, Message>() {

        @Override
        public void handle(Boot content, Message context) {
            if (state == State.WAITING) {
                LOG.info("{} Booting up.", self);
                trigger(new Booted(content.assignment), bootstrap);
                trigger(new CancelPeriodicTimeout(timeoutId), timer);
                trigger(new Message(self, server, Ready.event), net);
                state = State.STARTED;
            }
        }
    };

    @Override
    public void tearDown() {
        trigger(new CancelPeriodicTimeout(timeoutId), timer);
    }

    {
        subscribe(startHandler, control);
        subscribe(timeoutHandler, timer);
        subscribe(bootHandler, net);
    }
}

File: server/src/main/scala-2.11/bootstrap/BootstrapClient.scala
package bootstrap

import java.util.UUID

import bootstrap.ClientState.State
import com.typesafe.scalalogging.StrictLogging
import networking.{NetAddress, NetMessage}
import overlay.PartitionLookupTable
import se.sics.kompics.Start
import se.sics.kompics.network.Network
import se.sics.kompics.sl.{ComponentDefinition, NegativePort, _}
import se.sics.kompics.timer.{CancelPeriodicTimeout, SchedulePeriodicTimeout, Timer}

object ClientState extends Enumeration {
    type State = Value
    val Waiting, Started = Value
}

class BootstrapClient extends ComponentDefinition with StrictLogging {
    val bootstrap: NegativePort[Bootstrapping] = provides[Bootstrapping]

    val timer = requires[Timer]
    val network = requires[Network]

    val self: NetAddress = cfg.getValue[NetAddress]("stormy.address")
    val server: NetAddress = cfg.getValue[NetAddress]("stormy.coordinatorAddress")

    private var state: State = ClientState.Waiting
    private var timeoutId: Option[String] = None

    ctrl uponEvent {
        case _: Start => handle {
            logger.info(s"Starting bootstrap client on $self")
            val timeout: Long = cfg.getValue[Long]("stormy.keepAlivePeriod") * 2
            val spt = new SchedulePeriodicTimeout(timeout, timeout)
            spt.setTimeoutEvent(new BootstrapTimeout(spt))
            trigger(spt -> timer)
            timeoutId = Some(spt.getTimeoutEvent.getTimeoutId.toString)
        }
    }
    network uponEvent {
        case context@NetMessage(source, self,  Boot(assignment: PartitionLookupTable)) => handle {
            logger.info(s"Booting up $self")
            logger.debug(s"$context with $assignment")
            trigger(Booted(assignment) -> bootstrap)
            trigger(new CancelPeriodicTimeout(UUID.fromString(timeoutId.get)) -> timer)
            trigger(NetMessage(self, server, Ready) -> network)
            state = ClientState.Started
        }
    }

    timer uponEvent {
        case _: BootstrapTimeout => handle {
            state match {
                case ClientState.Waiting =>
                    trigger(NetMessage(self, server, CheckIn) -> network)
                case ClientState.Started =>
                    trigger(NetMessage(self, server, Ready) -> network)
                    suicide() // TODO Determine why we suicide here (and what it does)
            }
        }
    }

    override def tearDown(): Unit = {
        trigger(new CancelPeriodicTimeout(UUID.fromString(timeoutId.get)) -> timer)
    }
}
